pallets/moonbeam-foreign-assets/src/impl_fungibles.rs

```diff
@@ -0,0 +1,183 @@
+		let as_u128 = balance.saturated_into::<u128>();
+		Self::Balance::from(as_u128)
+	}
+
+	fn balance(asset: Self::AssetId, who: &T::AccountId) -> Self::Balance {
+		let balance =
+			EvmCaller::<T>::erc20_balance_of(asset, T::AccountIdToH160::convert(who.clone()))
+				.unwrap_or(U256::zero());
+		let as_u128 = balance.saturated_into::<u128>();
+		Self::Balance::from(as_u128)
+	}
+
+	fn reducible_balance(
+		asset: Self::AssetId,
+		who: &T::AccountId,
+		_preservation: frame_support::traits::tokens::Preservation,
+		_force: frame_support::traits::tokens::Fortitude,
+	) -> Self::Balance {
+		let balance =
+			EvmCaller::<T>::erc20_balance_of(asset, T::AccountIdToH160::convert(who.clone()))
+				.unwrap_or(U256::zero());
+		let as_u128 = balance.saturated_into::<u128>();
+		Self::Balance::from(as_u128)
+	}
+
+	fn can_deposit(
+		asset: Self::AssetId,
+		_who: &T::AccountId,
+		amount: Self::Balance,
+		provenance: Provenance,
+	) -> DepositConsequence {
+		let Some(location) = AssetsById::<T>::get(&asset) else {
+			return DepositConsequence::UnknownAsset;
+		};
+		let Some(asset_info) = AssetsByLocation::<T>::get(&location) else {
+			return DepositConsequence::UnknownAsset;
+		};
+		let status = asset_info.1;
+		// Check for total supply overflow
+		if provenance == Provenance::Minted {
+			let total_supply = EvmCaller::<T>::erc20_total_supply(asset).unwrap_or(U256::zero());
+			let minted_amount = U256::from(amount);
+			let Some(_new_total_supply) = total_supply.checked_add(minted_amount) else {
+				return DepositConsequence::Overflow;
+			};
+		};
+		match (status, provenance) {
+			(AssetStatus::FrozenXcmDepositForbidden, _) => DepositConsequence::Blocked,
+			(AssetStatus::FrozenXcmDepositAllowed, Provenance::Minted) => {
+				DepositConsequence::Success
+			}
+			(AssetStatus::Active, _) => DepositConsequence::Success,
+			(_, _) => DepositConsequence::Blocked,
+		}
+	}
+
+	fn can_withdraw(
+		asset: Self::AssetId,
+		who: &T::AccountId,
+		amount: Self::Balance,
+	) -> WithdrawConsequence<Self::Balance> {
+		if Self::asset_exists(asset) {
+			let balance = Self::balance(asset, who);
+			if balance >= Self::Balance::from(amount) {
+				WithdrawConsequence::Success
+			} else {
+				WithdrawConsequence::BalanceLow
+			}
+		} else {
+			WithdrawConsequence::UnknownAsset
+		}
+	}
+
+	fn asset_exists(asset: Self::AssetId) -> bool {
+		AssetsById::<T>::contains_key(&asset)
+	}
+}
+
+impl<T: Config> Create<T::AccountId> for Pallet<T> {
+	fn create(
+		_id: Self::AssetId,
+		_admin: T::AccountId,
+		_is_sufficient: bool,
+		_min_balance: Self::Balance,
+	) -> sp_runtime::DispatchResult {
+		sp_runtime::DispatchResult::Err(DispatchError::Other(
+			"Not implemented, must create through create_foreign_asset",
+		))
+	}
+}
+
+impl<T: Config> Unbalanced<T::AccountId> for Pallet<T> {
+	fn handle_dust(_dust: frame_support::traits::fungibles::Dust<T::AccountId, Self>) {}
+	fn set_total_issuance(_asset: Self::AssetId, _amount: Self::Balance) {}
```

Comment:
Maybe we can add a `log::warn` here, in `handle_dust`, and `set_total_issuance`

Reply:
I think we could add `debug` since it is normal for these functions to be called as part of the execution flow. wdyt?

Reply:
This would lead to many output logs since they are called in many functions of [`Mutate`](https://github.com/paritytech/polkadot-sdk/blob/c31ab2a246af0318a989ce6486f0bf8d7d385243/substrate/frame/support/src/traits/tokens/fungibles/regular.rs#L256) and other traits. I don't think we should add them.